opencv_ai/splice/io.h

```diff
@@ -0,0 +1,166 @@
+#ifndef OPENCV_AI_SPLICE_IO_H_
+#define OPENCV_AI_SPLICE_IO_H_
+
+#include <opencv_ai/splice/component.h>
+#include <boost/signals2.hpp>
+
+namespace opencv_ai {
+
+template <typename MessageType>
+class ChannelService : public boost::asio::io_service::service {
+ public:
+  ChannelService(boost::asio::io_service& io_service)
+      : boost::asio::io_service::service(io_service) {}
+
+  /// The unique service identifier.
+  static boost::asio::io_service::id id;
+  // Required by base class.
+  void shutdown_service() override {}
+
+  typedef boost::signals2::signal<void(MessageType)> Signal;
+
+  Signal& LookUpSignal(const std::string& channel_name) {
+    return signals_[channel_name];
+  }
+
+ private:
+  std::map<std::string, Signal> signals_;
+};
+
+template <typename MessageType>
+boost::asio::io_service::id ChannelService<MessageType>::id;
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+template <typename MessageType>
+class Output {
+ public:
+  typedef typename ChannelService<MessageType>::Signal Signal;
+
+  Output(const Context& context, std::string name)
+      : ctx_(context),
+        name_(std::move(name)),
+        full_name_(name_),
+        service_(boost::asio::use_service<ChannelService<MessageType>>(
+            ctx_.GetIoService())) {}
+
+  Output(const Component& component, std::string name)
+      : ctx_(component.GetContext()),
+        name_(std::move(name)),
+        full_name_(NameJoin(component.FullName(), name_)),
+        service_(boost::asio::use_service<ChannelService<MessageType>>(
+            ctx_.GetIoService())) {}
+
+  void Connect(std::string channel_name) {
+    CHECK(!Connected());
+    channel_name_ = std::move(channel_name);
+    signal_ = service_.LookUpSignal(channel_name_);
+  }
+
+  bool Connected() const { return !!signal_; }
+
+  std::string Name() const { return name_; }
+
+  std::string FullName() const { return full_name_; }
+
+  std::string ChannelName() const {
+    CHECK(Connected());
+    return channel_name_;
+  }
+
+  void Send(const MessageType& m) {
+    if (signal_) {
+      (*signal_)(m);
+    }
+  }
+
+ private:
+  Context ctx_;
+  std::string name_, full_name_;
```

Comment:
Looking at these it appeared that really they only depend on Component for the name, and that an Input or Output could exist with only a context.  Since this is inception, I haven't fully fleshed out what the purpose of Component is other than a way to namespace inputs and outputs...